rosbag2_transport/src/rosbag2_transport/player_service_client.cpp

```diff
@@ -0,0 +1,314 @@
+                "The configuration of introspection for '%s' is metadata on service side !",
+                service_name_.c_str());
+            }
+            break;
+          }
+        default:  // request_info_from::SERVICE:
+          {
+            // Already decide using request data from service.
+            ret = true;
+          }
+      }
+    } else {
+      if (service_set_introspection_content_) {
+        request_info_[client_id] = request_info_from::SERVICE;
+        ret = true;
+      } else {
+        request_info_[client_id] = request_info_from::NO_CONTENT;  // Only have metadata
+      }
+    }
+
+    return ret;
+  }
+
+  // type is service_msgs::msg::ServiceEventInfo::REQUEST_SENT
+  if (iter != request_info_.end()) {
+    switch (iter->second) {
+      case request_info_from::CLIENT:
+        {
+          // Already decide using request data from client.
+          ret = true;
+          break;
+        }
+      case request_info_from::NO_CONTENT:
+        {
+          if (rosbag2_cpp::service_event_include_metadata_and_contents(message.buffer_length)) {
+            // introspection type is changed from metadata to metadata + contents
+            request_info_[client_id] = request_info_from::CLIENT;
+            ret = true;
+          } else {
+            RCUTILS_LOG_WARN_ONCE_NAMED(
+              ROSBAG2_TRANSPORT_PACKAGE_NAME,
+              "The configuration of introspection for '%s' client [ID: %s]` is metadata !",
+              rosbag2_cpp::client_id_to_string(client_id).c_str(),
+              service_name_.c_str());
+          }
+          break;
+        }
+      default:  // request_info_from::SERVICE:
+        {
+          // Already decide using request data from service.
+          ret = false;
+        }
+    }
+  } else {
+    if (rosbag2_cpp::service_event_include_metadata_and_contents(message.buffer_length)) {
+      request_info_[client_id] = request_info_from::CLIENT;
+      ret = true;
+    } else {
+      request_info_[client_id] = request_info_from::NO_CONTENT;
+    }
+  }
+
+  return ret;
+}
+
+void PlayerServiceClient::async_send_request(const rcl_serialized_message_t & message)
+{
+  int ret = RMW_RET_OK;
+
+  {
+    auto ros_message = std::make_unique<uint8_t[]>(service_event_members_->size_of_);
+
+    service_event_members_->init_function(
+      ros_message.get(), rosidl_runtime_cpp::MessageInitialization::ZERO);
+
+    ret = rmw_deserialize(
+      &message, service_event_type_ts_, ros_message.get());
+    if (ret == RMW_RET_OK) {
+      if (client_->service_is_ready()) {
+        // members_[0]: info, members_[1]: request, members_[2]: response
+        auto request_offset = service_event_members_->members_[1].offset_;
+        auto request_addr = reinterpret_cast<size_t>(ros_message.get()) + request_offset;
+        auto future_and_request_id = client_->async_send_request(
+          reinterpret_cast<void *>(*reinterpret_cast<size_t *>(request_addr)));
```

Comment:
@Barry-Xu-2018 Could you please explain this pointers arithmetic?
In particular it is unclear to me why need to cast pointer to the `uint8_t` to the `size_t` before adding offset here `reinterpret_cast<size_t>(ros_message.get()) + request_offset;`?

In my understanding we shall not cast it to the size_t if `request_offset` in bytes and we have a pointer in memory with type of the `uint8_t *`. Adding number of bytes to those pointer shall give us a new pointer to the required offset.
i.e. 
```
void * request_ptr = ros_message.get() + request_offset;
```
should do the math for the new pointer. ??
Or `request_offset` is not in bytes but rather in 4 bytes quantities?

Reply:
Ahh never mind. I've figured out. The trick is that  `ros_message.get() + request_offset;` give us a pointer to the bounded sequence of the requests and we need to get the first element from it.
I will rework this part.